sw/Core/Src/touch.h

```diff
@@ -545,148 +547,338 @@ s16 midi_chan_pitchbend[16];
+	//
+	// u8 numfingerdown = 0;
+	// for (int attempt = 0; attempt < 16; ++attempt) {
+	// 	u8 ch = (midi_next_finger++) & 7;
+	// 	bool fingerdown = touch_synth_getlatest(ch)->pressure > 0 && (midi_pressure_override&(1<<ch))==0; // synth_dst_finger>0, and midi_pressure_override bit is not set
+	// 	if (fingerdown) {
+	// 		if (attempt < 8) {
+	// 			numfingerdown++;
+	// 			if (numfingerdown == 8)
+	// 				return 255; // all fingers using channels! NO ROOM! TOUGH
+	// 		}
+	// 		continue;
+	// 	}
+	// 	if (attempt >=8 || midi_channels[ch] == 255)
+	// 		return ch;
+	// }
+	// // give up
+	// return (midi_next_finger++) & 7;
 }
 
 bool is_finger_an_edit_operation(int fi);
 
 void finger_synth_update(int fi) {
-	if (fi == 0) {
-		total_ui_pressure = 0;
-	}
+	static u8 last_edited_step_global = 255;
+	static u8 last_edited_step[8] = {255, 255, 255, 255, 255, 255, 255, 255};
+	static u8 last_edited_substep[8] = {255, 255, 255, 255, 255, 255, 255, 255};
+	static u8 step_full = 0;
+
 	int bit = 1 << fi;
-	int frame_ui = (finger_frame_ui-1) & 7;
-	if (finger_ui_done_this_frame & bit)
-		frame_ui = (frame_ui + 1) & 7;
-//	int prevframe_ui = (frame_ui - 1) & 7;
-	int prevprevframe_ui = (frame_ui - 2) & 7;
-//	int prev4frame_ui = (frame_ui - 4) & 7;
-	Finger* uif = &fingers_ui_time[fi][frame_ui];
-//	Finger* uif_prev = &fingers_ui_time[fi][prevframe_ui];
-	Finger* uif_prevprev = &fingers_ui_time[fi][prevprevframe_ui];
-//	Finger* uif_prev4= &fingers_ui_time[fi][prev4frame_ui];
-	Finger* synth_dst_finger = &fingers_synth_time[fi][finger_frame_synth];
-//	Finger* synth_prev_finger = &fingers_synth_time[fi][(finger_frame_synth-1)&7];
-
-	int p = uif->pressure;
-//	int prevp = uif_prev->pressure;
-	int prevprevp = uif_prevprev->pressure;
-//	int prev4p = uif_prev4->pressure;
-//	bool fingerup = (p <= 0 && prevprevp <= 0 && prev4p<=0);
-	bool pressure_not_rapidly_decreasing = (p > (prevprevp)); // we do this BEFORE compensating for edit mode, so that we dont accidentally see it as increasing after edit ends
-	if (editmode != EM_SAMPLE) {
-		if ((fingerediting | prevfingerediting) & bit)
-			p = -256;
-		else if (shift_down == SB_CLEAR)
-			p = -256;
-		else if (is_finger_an_edit_operation(fi))
-			p = -256;
+	int ui_frame = (finger_frame_ui - (finger_ui_done_this_frame & bit ? 0 : 1)) & 7;
+	Finger* ui_finger = &fingers_ui_time[fi][ui_frame];
+	Finger* synth_finger = &fingers_synth_time[fi][finger_frame_synth];
+	int previous_pressure = fingers_ui_time[fi][(ui_frame - 2) & 7].pressure;
+	int previous_position = fingers_ui_time[fi][(ui_frame - 2) & 7].pos;
+	int substep = calcseqsubstep(0, 8);
+	bool latchon = (rampreset.flags & FLAGS_LATCH);
+	int pressure;
+	int position = ui_finger->pos;
+	bool pressure_increasing;
+	bool ignore_touch_for_synth = false;
+
+	// midi and finger-touches are handled on a first come, first serve basis
+	// if this string is already playing a midi note, the finger touch is ignored
```

Comment:
this is the only part of this that I'm said about BUT I appreciate it is a step on the path to victory, so we can take it for now! just wanted to note that this is room for future improvement